### ISSUE TITLE
Modern System Tray Icon

### ISSUE THREAD

**Issue by mveril** [microsoft/ProjectReunion#713]
# Proposal: Modern System Tray Icon

## Summary

After a feature proposal to the WinUI team to make system tray icons with WInUI microsoft/microsoft-ui-xaml#2020 and to have followed the following discussions #519. and proposals #711  I would like to offer my take on what the modern way to create system tray icons should look like. A system-managed tray icon accessible for UWP and WIn32 applications respecting the app lifecycle #111  and offering a UI that respects WinUI standards.

## Rationale

* The system tray icons are not dead and still used a lot.
* A refresh of the UI is necessary to follow the system theme (not the application theme in this case) and to use WinUI for example for the contextual menu. (Like the One drive system tray icon).
* The actual System tray API needs an app that runs all the time but the management of the context menu and user actions can be handled by the system and through background tasks.

## Scope

| Capability                                                   | Priority |
| :----------------------------------------------------------- | :------- |
| This proposal will allow UWP and Win32 developers to create System tray icon that respect the app lifecylce | Must     |
| This proposal will allow UWP and Win32 developers to create System tray icon that respect Windows theme and fluent design | Must     |
| This proposal will allow UWP and Win32 developers to modify his System tray icon from background and foreground | Must     |
| This proposal will allow end users to run background tasks using the system tray icon | Must     |
| This proposal will allow end users to control background tasks using the system tray icon | Should   |
| This proposal will allow developers to show flyout using the system tray Icon | Could    |
| This proposal will allow UWP and Win32 developers to show balloon notifications | Won't    |

## Important Notes

[contactPanel API](https://docs.microsoft.com/windows/uwp/contacts-and-calendar/my-people-support) could be a good inspiration for this new tray icon API espescialy for the flyouts.

**Comment by orcmid** [comment]
As an user, I find the increasing use of system tray icons and keeping apps running in the background to be quite offensive.  In particular, too many apps force going to the system tray on close with no user control of the behavior.

 1. Zoom on Windows 10 does it better - it ends up creating a task bar icon that reveals Zoom is still running, and while objectionable, it is easy to notice and to exit it.
 2. My Philips DreamWeaver data card loader, an app I only run once a week, goes to the system tray whether I like it or not (although there is a notification that happened).  Then I Have to go kill it.  There is absolutely no reason for this application to stay running in the background.  And I still use the removable drive removal to ensure that the USB key from my C-PAP can be safely removed.  
 3. The change of the Bing Desktop App to just be a Bing Wallpaper app that sits in the system tray is very annoying.  That is the only way, now, to find out what the wallpaper image subject is and where to find more information about it.  That's a tiny symbol and I have to perform extra work.  It keeps me from doing more Bing searches every day.  I doubt that shows up in their stats, but I use Bing search less than before.
 4.  The Discord app is another offender.  I learn that I have left it running by virtue of all the notifications that pop up, until I go into the system tray pop-up and kill it dead.
 5. What makes this even more difficult is apparent absence of a way to force certain apps to have their system tray icon be visible and not hidden in the pop=up of all of them.  I made good use of that provision in the past.  I can't find it on the currently-supported Windows 10 releases.

**Comment by ptorr-msft** [comment]
@orcmid presumably a new API would solve those problems:
1. Users must grant permission for the icon to exist, with a mechanism to manage them after-the-fact, too.
2. The target app is suspended / terminated, and only woken up when the user interacts with the icon.

**Comment by mveril** [comment]
> @orcmid presumably a new API would solve those problems:
Yes I think
> 1. Users must grant permission for the icon to exist, with a mechanism to manage them after-the-fact, too.
Yes but I don't know if Win32 app should be allowed to enable it by default like startup action or not
> 2. The target app is suspended / terminated, and only woken up when the user interacts with the icon.
Exactly It's the main interest of this proposal. It's what I called "respect the app lifecycle"

**Comment by huoyaoyuan** [comment]
System tray icon can also (and commonly) be used for an interface to service-like applications. The application runs background tasks and the icon is active.

**Comment by orcmid** [comment]
> System tray icon can also (and commonly) be used for an interface to service-like applications. The application runs background tasks and the icon is active.

Yes, and I see those.  They are not the subject of my complaint.  

I am a little bemused that I am not explicitly running any graphical programs but 15 are using my GTX 1060 6GB display adapter.  That's an usefull systrap thingie.  I am also happy with my password monitor and HyperSnap being there and on the displayed system tray.  

I am about to uninstall Skype since most of those calls are now accomplished with Microsoft Teams.  Neither Teams nor the "Your Phone" app seem to need residence on the system tray.

I am not so thrilled with Outlook,  OneNote, and WinZip taking system folder real-estate though.  I'm going to dump the Wnzip and Outlook ones.

**Comment by riverar** [comment]
> The target app is suspended / terminated, and only woken up when the user interacts with the icon.

Our app consists of a flyout that appears when a user clicks a notification area icon. We cannot afford to be suspended and resumed, it must remain running in the background.

On those grounds, I reject this proposal.

**Comment by ptorr-msft** [comment]
@riverar remember this is Reunion - it's all about options. I should have been clearer in my reply.

The actual requirement listed above is:

> This proposal will allow UWP and Win32 developers to create System tray icon that respect the app lifecylce

"respect[ing] app lifecycle" means the app might suspend (if it wants to) or it might not. The problem with `Shell_NotifyIcon` today is that the app *MUST* keep running to check window messages; that's a poor design if your app doesn't need to be running 24x7.

**Comment by riverar** [comment]
@ptorr-msft Hm, I don't agree. @mveril states in their summary (emphasis mine):

> A system-managed tray icon accessible **for UWP and WIn32 [sic] applications respecting the app lifecycle** #111 and offering a UI that respects WinUI standards.

The author further lists the following as a **Must**:
> This proposal will allow UWP and Win32 developers to create System tray icon that respect the app lifecylce [sic]

While the author wrote "will allow", I believe that is in error. It appears the author's intention is to _require_ apps to participate in PLM to take advantage of the proposed enhancements. I don't think that's fair to apps, like ours, that cannot participate in PLM.

Suggested scope changes:
```diff
- This proposal will allow UWP and Win32 developers to create System tray icon that respect the app lifecylce | Must
+ New mechanism will allow apps participating in PLM to create a modern Notification Area icon | Must
+ New mechanism will allow apps not participating in PLM to create a modern Notification Area icon | Must
```

**Comment by ptorr-msft** [comment]
If you read the spec for #111, nothing about suspension is mandated. It talks about "allowing" apps to be "notified" which implies they can choose to ignore the notification (or not even register for it). There is no forced "lifecycle" in Reunion, so if @mveril **wants** this proposal to require suspension, then a different change would be needed to the requirements.

**Comment by mveril** [comment]
@riverar Sorry I haven't had time to test your app yet, but if I understand your needs correctly, IMHO it won't be necessary to have an always-on app to display a flyout on click. I don't know if you've paid attention to this but this proposal will allow support for flyouts with an API that might look like the [Contact panel](https://docs.microsoft.com/windows/uwp/contacts-and-calendar/my-people-support#running-in-the-contact-panel)
I understand that a lot of apps can't comply with PLM.  (Docker, Antivirus and many more ...).  You're right, more flexibility is always better, this is "Project Reunion" after all!
  So what would you imagine from an implementation perspective?  Some APIs for app which support PLM with Background tasks and activation events and an alternative in the form of classical events to leave the choice, like the `Window` class of WinUI which supports both CoreWindow and HWind with certain APIs which are different for each case, or will a single API surface be possible to handle both cases?

**Comment by mveril** [comment]
@riverar I finally had the time to test your application (very good job ! I think I'll keep it on my system). Here is my conclusion:
> [Our app](https://www.microsoft.com/en-us/p/eartrumpet/9nblggh516xp) consists of a flyout that appears when a user clicks a notification area icon. [...] it must remain running in the background.

With this proposal, it will not be necessary (but must be possible) to have an application always active to display a flyout on click. this functionality should be handled by this proposal (because It is widely used) and the flyout would be created for you. your job would only be to define the content of the flyout menu and adjust the size.
This corresponds to the scope:
> This proposal will allow developers to show flyout using the system tray Icon

But in your particular case even if I have no certainty I think that you will still need an always active application to manage the global hot keys but that should absolutely not prevent you from using this feature proposal properly

**Comment by ptorr-msft** [comment]
I think the current wording is fine, but you can change if you like (it's your proposal :)).

The problem I believe you want to avoid (I sure do!) is the current requirement for apps to always be running to respond to the tray via the WndProc. 

There are two obvious ways to improve on this:

- Integrate with the Reunion suspension model (whatever that is :)).
- Relaunch the app with appropriate context.

Both should be available in such an API.

**Comment by mveril** [comment]
@ptorr-msft Yes, that's exactly the goal, but this new API should also be easily accessible for long-running service-like applications like @huoyaoyuan as previously pointed out.
Maybe the [AppInstance ](https://github.com/microsoft/ProjectReunion/blob/main/specs/AppLifecycle/Instancing/AppLifecycle%20SingleMulti-Instancing.md) ProjectReunion API proposed by @andreww-msft as part of #111  should be the solution.
Anyway I don't think an application using a System tray icon will have any interest in displaying an icon per instance. Either the application will be multi-instance, the icon will be shared.
Otherwise the activation event could refer to the instance which created the icon in its properties.
What do you think of this solution ?

**Comment by orcmid** [comment]
> What do you think of this solution ?

I am getting the impression that this situation is being over-thought and is maybe over-constrained as well.

Maybe it would be good to elevate from implementation worries to an agreement on the function and utility of the system tray for users, accounting for the degree of Windows using population that has no understanding of it and blissfully ignore notifications too.

As I've already mentioned, I give Zoom great credit for using the task bar to reveal it is running minimized on "exut,"  That Discord hides in the system tray, usually out of sight, is very annoying and revealed only when there are notifications about activity in which I have no interest.  So I guess another higher level consideration has to do with how products might abuse the tray and also its employment for surveillance actions.

**Comment by orcmid** [comment]
> ... Discord hides in the system tray, usually out of sight, is very annoying and revealed only when there are notifications about activity in which I have no interest. I know to hunt it down now, but it is making me work too hard. 

Part of my dismay is that the ability to control what appears in the system tray and system-tray strip od the task bar has disappeared.  Duhhh.  It turns out that the "system tray" name is not used and the Settings / Personalization / Taskbar / Notification area / Select which icons appear on the taskbar is what I needed instead.  Sheesh.  OK, working for me.  So maybe that is also what we should be calling it and also comprehending the intended use case.

**Comment by mveril** [comment]
@orcmid  sorry for not having answered you earlier but I was very busy with my job this week.
You are right, it is important to leave the maximum number of doors open while you find the right scopes.
Personally, I don't really like the term 'Notification area' because today balloon notifications are deprecated and replaced by toast notifications. therefore the icons of the system tray no longer fulfill this purpose.
Personally, I don't think the main taskbar part is the best place to monitor or report background tasks. If I understood correctly what you said, this is what Zoom does? for me this part is reserved for the foreground apps and the favorite app.
I think that a specific authorization should be requested by the application to display in the system tray as is the case for startup tasks. These tasks could be a good model for the authorization to display its icons.
The peculiarity is that the startup tasks does not display a confirmation request for the Win32 model unlike the UWP model.
This authorization could be accepted or refused at any time in the settings with the section you are talking about but also in the parameters of each application as can be done with the start-up tasks.
Should we ask permission explicitly, for which app model, UWP or all?
Do you think these icons should be completely non-existent when it is off or appearing in the drop-down menu as it is now?
Maybe both cases could be handled.
This API could also be linked to a specific permission in the manifest to limit abuse.
We can also think of a menu item called "Hide" which will allow the user to easily send the icons to the drop-down menu and which will automatically be displayed at the bottom of the contextual menu.
What will be your wishes for the system tray?

**Comment by sylveon** [comment]
@orcmid you can also just drag and drop icons to and from the tray overflow instead of going in settings ;)

On to the issue.

I've recently commited work to integrate a UWP MenuFlyout with Shell_NotifyIcon and while it's a bit buggy (in part due to https://github.com/microsoft/microsoft-ui-xaml/issues/3953) it looks pretty solid and modern:

![https://cdn.discordapp.com/attachments/660265525124923392/835385558146482196/ttb_rocks.mp4](https://cdn.discordapp.com/attachments/660265525124923392/835385558146482196/ttb_rocks.mp4)

I suspect one could use any flyout as well, and I plan on trying that in future versions for a better UX and less crashes.

My app is "service-like" so does need 24/7 execution, and I would be interested in replacing the 2000 lines of code this required for something more... Simple to implement.

A few requirements I can think of:

- support for any flyout, be it MenuFlyout or just a Flyout with custom controls.
- a modern looking tooltip, my implementation does that too :)
- as mentioned before, does not require processes to opt-in to lifecycle
- can be strictly bound to a specific application, like Shell_NotifyIcon with a GUID can, but without the requirement to code sign the executable. This has the bonus side effect that if a background app crashes and restarts, the tray icon isn't duplicated ad infinitum
- support for keyboard only interaction by default (tab navigation from the taskbar to the tray icon and hitting space/enter should open the flyout). While implementing this one thing I noticed is that the amount of apps doing this was... Low. I implemented it because it was a few lines of code more but I can see why it's easy to omit it.

**Comment by mveril** [comment]
> I've recently commited work to integrate a UWP MenuFlyout with Shell_NotifyIcon and while it's a bit buggy (in part due to [microsoft/microsoft-ui-xaml#3953](https://github.com/microsoft/microsoft-ui-xaml/issues/3953)) it looks pretty solid and modern:
> 
> https://cdn.discordapp.com/attachments/660265525124923392/835385558146482196/ttb_rocks.mp4
> 
> I suspect one could use any flyout as well, and I plan on trying that in future versions for a better UX and less crashes.

Very good job @sylveon  !
From a UX point of view, this is exactly the goal.
Is the app you show in the video open-source? I will be interested to look at the code and test it.
> 
> My app is "service-like" so does need 24/7 execution, and I would be interested in replacing the 2000 lines of code this required for something more... Simple to implement.
> 
> A few requirements I can think of:
> 
> * support for any flyout, be it MenuFlyout or just a Flyout with custom controls.

It's planned in the scope
In fact, there are generally two types of interface:

- A context menu (displayed to the user as a [MenuFlyout ](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.menuflyout), generaly on right click.

- A [Flyout ](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.flyout) that can be filled by any control, generaly on left click.

> * a modern looking tooltip, my implementation does that too :)

Yes, I think the Fluent design must be applied at all levels
> * as mentioned before, does not require processes to opt-in to lifecycle

Yes, this case must be fully supported. Maybe using COM or with some specific APIs.

> * can be strictly bound to a specific application, like Shell_NotifyIcon with a GUID can, but without the requirement to code sign the executable. This has the bonus side effect that if a background app crashes and restarts, the tray icon isn't duplicated ad infinitum

Oh I didn't know it's possible to do that! This is a problem that I have encountered before. This would be even more essential since the icon should be managed by the system and could be eventually used by several successive instances of the same app.
> * support for keyboard only interaction by default (tab navigation from the taskbar to the tray icon and hitting space/enter should open the flyout). While implementing this one thing I noticed is that the amount of apps doing this was... Low. I implemented it because it was a few lines of code more but I can see why it's easy to omit it.

You are right I didn't know it was already possible. 
An out of the box standardization of keyboard shortcuts would be interesting.

I have updated the scope to reflect your comment. thank you very much ! I'm very happy to see it's possible using Shell_NotifyIcon APIs to create Fluent System tray icon UX with WinUI. This shows the way !

**Comment by sylveon** [comment]
Yes it's open source! The code is in [this commit](https://github.com/TranslucentTB/TranslucentTB/commit/7f826762ab0358ff3c5e5d0a2241dafd67fe3927) and you can grab a build to test from [Azure Pipelines](https://dev.azure.com/sylve0n/TranslucentTB/_build?definitionId=4)! The code is licensed under GPLv3, but I don't particularly mind anybody taking inspiration from it.

You'll have to add the following in the configuration file to enable it, since it's still crashy I'm leaving it undocumented and off by default for now:
```json
"experimental": {
    "use_xaml_context_menu": true
}
```

**Comment by sylveon** [comment]
In Windows 11, custom tooltips for tray icons seems to be intentionally disabled.

They all look like this, even if your app says it wants to show custom tooltips the taskbar will disregard it:
![image](https://user-images.githubusercontent.com/6440374/126467000-fe14b3dc-478a-4dd8-99b6-f51def9def4f.png)

VS in Windows 10:
![image](https://user-images.githubusercontent.com/6440374/126467035-fc560e61-c434-41e5-af9d-cde4b28329ff.png)

I left feedback about it here: https://aka.ms/AAd21h2

**Comment by zadjii-msft** [comment]
> In Windows 11, custom tooltips for tray icons seems to be intentionally disabled.

I'm pretty sure this thread isn't the place to follow up with that particular bug, but I'm not totally sure how the WindowsAppSDK team wants to triage their repo. For other internal folks reading this thread, that feedback got parented to MSFT:34113237.

**Comment by sylveon** [comment]
Was just pointing it out because I've previously talked about my implementation using custom tooltips. Glad to know there's an internal bug for it though!

**Comment by Jaiganeshkumaran** [comment]
One of the problems with notification area icons are that it's primarily an app powered experience which means it needs to run all the time. While many of them are to indicate they are running in the background, some of them just use it to display options. Jumplists could technically be used for a lot of them however now with Windows 11 not showing them in Start, there's less place to access them and jumplists don't indicate an app running in background or not. Notification area icons can continue to be used however instead of apps displaying them, it should be brokered and the options in the context menu, action to do when clicked etc... should be well defined and instead of being part of an UI framework, it should be part of the application model itself. Since the apps only say what options and don't just attach their own, Windows can use a more modern UI and update anytime.

**Comment by Jaiganeshkumaran** [comment]
With jumplists, it's not possible to change options dynamically unless you're running all the time. Maybe there could be a way where a background task that updates the options for the current program state.

**Comment by Jaiganeshkumaran** [comment]
System tray icons shouldn't be used if there's no background task or need for quick access to options even without the app opened that refresh dynamically. Some apps like Spotify uses a tray icon as a way to minimize the application when playing music which shouldn't be done as the user doesn't know whether a foreground UI is actually there and will wonder where the sound is coming from. If an app shows a tray icon, it should be shown even without the foreground UI and shouldn't be a companion to it.

**Comment by HavenDV** [comment]
I want to note for the future that all this is already implemented in my [H.NotifyIcon](https://github.com/HavenDV/H.NotifyIcon) library. It also has an example of a Windowless application, as well as an example of a regular WinUI application with the ability to collapse. It also supports the ability to create a native context menu for TrayIcon with animation (still in preview mode) and some other features. The API is accessible from xaml and mostly inherited from [hardcodet/wpf-notifyicon](https://github.com/hardcodet/wpf-notifyicon) for easy transition to WinUI from WPF

**Comment by riverar** [comment]
I would certainly hope not. Widgets are not a good substitute for notification area icons. For example, they don't even appear on-screen without opening the Widget Board.

**Comment by niels9001** [comment]
Beyond context menu's and icons, another use case is to show a flyout that would let you host a window.

For example, for PowerToys, we have a single click flyout that looks like this:

<img width="534" height="590" alt="Image" src="https://github.com/user-attachments/assets/84084ff7-1940-43f7-987c-e3d4277ae6b6" />

**Comment by sylveon** [comment]
With my strategy I already can show one of those if I want. After all a MenuFlyout is just another kind of Flyout.

**Comment by dotMorten** [comment]
I think the cleanest API for presenting a context menu would be to expose a `FlyoutBase` property. That'll give you a quick way to create a context menu using `MenuFlyout`/`CommandBarFlyout`, or if you want custom content, use the generic `Flyout`.

**Comment by riverar** [comment]
**Not a blocker** but something to throw out there: Beware that showing flyouts and other interactive windows _from the notification area_ will be subject to z-band ordering rules and quirks. This will, unfortunately, mean flyouts will be undismiss-able and/or present other issues in scenarios such as "Start is open" and full-screen games.

**Comment by mdtauk** [comment]
A common surface/window type that is similar to the Thin Acrylic the Widgets dashboard used to use, and a mica window without titlebar, would allow third party taskbar UI to better fit in with Windows.  First and Second-party Microsoft apps which have tray UI like OneDrive or PowerToys, could also have UI that uses WinUI and match the system, without having to roll their own bespoke thing.

Common API to handle user choice in disabling, hiding, showing, these tray icons, so good actors can thrive, and bad actors should be encouraged to conform, before new versions of Windows, make breaking changes for background apps which don't target the new APIs.

Context Menus and XAML UI that respect the chosen System Mode, rather than the App Mode.  As the UI is activated by the Taskbar which is the shell, and should be seen as part of the shell.  Enable Dark Mode Win32 support to better match the shell.

Behaviour to allow or disallow these tray apps for the Gaming SKUs of Windows coming for future handhelds and xbox devices.

APIs to distinguish between tray apps that are actively doing an action, those just inactive, and disabled ones perhaps.  APIs for the tray app process, as well as the App which installed the tray app - so apps can know when tray apps are disabled.

**Comment by riverar** [comment]
Calls for greater shell extensibility are nearly as old as the Windows shell itself. To temper expectations, I don't think we'll see much movement on the addition of any further extensibility or native surfaces. Last I heard, Start and Taskbar have mostly been moved overseas for maintenance and servicing activities.

**Comment by sylveon** [comment]
As far as I understand, the system tray is mostly considered a legacy thing and modern apps are encouraged to use other surfaces like jump lists instead.

**Comment by dotMorten** [comment]
@sylveon That's the story the Windows team tried to tell during the UWP days that no one really bought into, and even new Microsoft apps are heavily relying on it, and for good reason. It's quite usable, and I think it's fair to say that developers are wanting to have the same ability using an easy-to-use API.

**Comment by svrooij** [comment]
@niels9001 thanks for re-igniting this issue, at least now there is some activity. And we still want this.

**Comment by mdtauk** [comment]
> Calls for greater shell extensibility are nearly as old as the Windows shell itself. To temper expectations, I don't think we'll see much movement on the addition of any further extensibility or native surfaces. Last I heard, Start and Taskbar have mostly been moved overseas for maintenance and servicing activities.

Is this the Windows 10 versions, or also for current versions of Windows 11?

**Comment by dotMorten** [comment]
For those who can't wait, I added a simple TrayIcon API to WinUIEx that allows you to add tray icons and use the WinUI FlyoutBase classes for interacting with them: https://dotmorten.github.io/WinUIEx/concepts/TrayIcon.html  `</ShamelessPlug>`

**Comment by aepot** [comment]
@dotMorten Thanks for the awesome `TrayIcon`.

But a small feedback: When the icon is enough high above screen bottom, then the Flyout pops above the icon and totally covers it. That's somewhat unexpected.

Probably the anchor may be at the mouse position not at the left+top corner of the Icon, or at the center of the icon.

<img width="185" height="162" alt="Image" src="https://github.com/user-attachments/assets/34b4901e-9289-44f7-99c0-2798efa0c373" />

The actual icon placement is highlighted with red square. There are two rows of the hidden Tray icons in the Windows native popup.

---

About expectations from the Tray Icon feature: I'm in migration from the WPF and my Product Team wants me to reproduce everything that the old app does but in brand-new Fluent Design skin with native Notification features, nice Backdrops, etc. The Tray Icon is mandatory here. @dotMorten's API pretty matches our expectations, special thanks for the `TrayIcon` flexible class API that allows Window lifetime to be split from the Icon's one. I suggest to take this to 2.x as it looks not that heavy. Please :)

**Comment by sylveon** [comment]
This is how all tray icons behave. For example here Spotify:

<img width="539" height="258" alt="Image" src="https://github.com/user-attachments/assets/7d4415c8-8307-4739-9a67-bb68b1ef6825" />

**Comment by aepot** [comment]
@sylveon correct but it repsects mouse pointer position, right?

**Comment by sylveon** [comment]
Yeah, it's anchored to pointer position.

**Comment by dotMorten** [comment]
@aepot I'd suggest logging this as feedback in the winuiex repo, since that's a bit off topic for this thread.

**Comment by aepot** [comment]
@dotMorten I guessed that some ideas implemented there may be used as a prototype for Windows App SDK. So that's my suggestion to pay attention to that detail while implementing it here. Thus it wasn't completely off topic.

BTW, it's up to you how to improve WinUIEx though. It's open-source and I can borrow the code to make some changes to it directly in my Project if something important is missing. Great thanks for the really good solution 👍

**Comment by inoyakaigor** [comment]
With an AI after 20 years in the web I went back to making desktop apps as pet projects. Back then I used to Borland С++ Builder 6 and if you know it has beatiful set of built-in components. One of them was [TTrayIcon](https://docwiki.embarcadero.com/CodeExamples/Athens/en/TTrayIcon_(C%2B%2B)) which was easy to use.

I'd like to have the same component.

**Comment by hoshiizumiya** [comment]
<img width="179" height="317" alt="Image" src="https://github.com/user-attachments/assets/881a9d5e-8240-46ec-b2a0-1e96c63546c6" />

<img width="269" height="226" alt="Image" src="https://github.com/user-attachments/assets/61ac5f3b-09c6-4880-a20b-5429227a76f7" />
Do it yourself. The Windows "AI" sdk will never add this.

**Comment by huoyaoyuan** [comment]
FWIW, it's mentioned in this [Build session](https://build.microsoft.com/en-US/sessions/OD854) at 3:32. At least the team is admitting this feature request.
